### ISSUE TITLE
SCALE 23.10.1 preliminary test issues

### ISSUE THREAD

**Issue by SayHeading** [ngandrass/truenas-spindown-timer#29]
Aware that spindown_timer.sh is not confirmed to be compatible with TrueNAS SCALE 23.10.1 but I tried and (so far…) failed.

Some observations:

(4 disks installed; 3*data (tank/SATA HDD) + 1*system SSD (boot-pool with system dataset/SATA SSD on USB))

Using disk names in automatic mode (e.g. `-i sdd`; where `sda, sdb, sdc` are data disks and `sdd` is the boot-pool) is problematic, as I observed that a reboot of TreuNAS SCALE sometimes messed up the disk names, here the boot-pool was observed to be `sdd` or `sdc` sometimes. I assume that this might be caused by how the disks are activated during boot and could probably be solved by using names iso identifiers for the spindown_timer.sh command.

But automatic mode `-i boot-pool` resulted in a report stating that the boot-pool was still monitored. 
`Monitoring drives with a timeout of 900 seconds: sda sdb sdc sdd`.

Manual mode `-m -i tank` seems to do the trick.
`Monitoring drives with a timeout of 900 seconds: tank`

But in the end spindown/shutdown is not yet working here, I might mess up the correct path to the script, as it now resides in `/home/admin/scripts` as you can see on the attached terminal screenshot.
The difference to CORE and it’s file directory might be my issue (and my innocence of it…).

I used `/home/admin/scripts/spindown_timer.sh -t 900 -p 600 -m -i tank -s 1800` as command. Would that be okay?
Or does the message `[ERROR] : No supported disk parameter tool found.` point towards something?

And then finally, just discovered (while typing this text) during the manually (via ssh) started script
`./spindown_timer.sh: line 586: shutdown: command not found` 
as you also can see in the screenshot.

![Bildschirmfoto 2024-01-13 um 14 50 52](https://github.com/ngandrass/truenas-spindown-timer/assets/156519194/2796c21c-859e-4056-86ca-568a3d32d640)

**Comment by ngandrass** [comment]
Thanks for your detailed report :+1: 

> I used /home/admin/scripts/spindown_timer.sh -t 900 -p 600 -m -i tank -s 1800 as command. Would that be okay?

Your command looks fine, though I'd set the idle timeout (`-t`) to the same value as your shutdown timer (`-s`) since you are shutting down the system after just 30 minutes of idling anyway. But that's just personal preference :)

> And then finally, just discovered (while typing this text) during the manually (via ssh) started script
./spindown_timer.sh: line 586: shutdown: command not found

This shows the problem pretty well: Please run the script as `root`. It looks like your `admin` user does not have the required privileges to perform spin down or shutdown operations ;)

**Comment by SayHeading** [comment]
> (...)I'd set the idle timeout (...). But that's just personal preference :)

Ah, these are only settings for the evaluation phase, not serious figures :-)

> This shows the problem pretty well: Please run the script as `root`. It looks like your `admin` user does not have the required privileges to perform spin down or shutdown operations ;)

Okay, so after I enabled `root` to have password access and ssh (what Cobia doesn´t allow as default) manually starting the script via terminal seems to work, but there is now a new issue:


![Bildschirmfoto 2024-01-14 um 18 52 39](https://github.com/ngandrass/truenas-spindown-timer/assets/156519194/9044865e-3831-4c13-8835-3f154a8a4576)

Looks like `tank` is not sufficient to spin all 3 drives down. But I´ll give it another trial with drive names and the boot-pool connected to SATA directly to have stable numeration.

But what is more important: When using the GUI to implement the script, does it matter who does it (logged in) , `admin` or `root`? As from the interface it doesn't give me other options to do so. Or is there something that has to be added to the GUI command to clarify it has to be applied with root privileges?

**Comment by SayHeading** [comment]
...first update, boot device now directly attached to SATA, using disk names as displayed in TrueNAS GUI (aka identifier) in auto mode.
Running script via terminal as root (disabled script in GUI).
But still no joy, shutdown: `invalid option -- 'p'`

![Bildschirmfoto 2024-01-15 um 11 47 27](https://github.com/ngandrass/truenas-spindown-timer/assets/156519194/8019117e-3e6e-4b2a-9e42-9bfa4976335d)

And still - although with the boot device now on SATA directly iso USB to SATA enclosure - do the device identifier/names change from reboot to reboot. Not a script issue, but one for the script user who monitors devices by identifier.

**Comment by Salvoxia** [comment]
Yes, the Debian `shutdown` command does not recognize the `-p` argument as FreeBSD does. The same result (Halt or Power off after shutdown) can be reached with `-h` under Debian based systems. The script would somehow need to distinguish between these two.

The disk device names changing between boots is a known behavior in TrueNAS Scale. AFAIK nothing you can do about that.

Side question because I have problems with that: If a drive is spun down for you, does it stay that way for a prolonged period of time without waking up "on its own"?

**Comment by SayHeading** [comment]
> The disk device names changing between boots is a known behavior in TrueNAS Scale. AFAIK nothing you can do about that.


Good to know it's not my hardware so I can revert using an USB->SATA attached boot drive to save SATA ports if needed.
And using `-u zpool` instead is a usable option, but it would be worth emphasizing the possible issue in the user instructions.

> Side question because I have problems with that: If a drive is spun down for you, does it stay that way for a prolonged period of time without waking up "on its own"?

Wasn't able to test that yet in prolonged period, as our electricity will be shut down soon during the afternoon. Up to know I just have a short time window of 30 minutes to report, where the disks were sleeping before I now have to shut down the system. I'll try again once power is up again.

In the meanwhile - but that is now explained by your comment - I could report that adding the script via the GUI as `root` spun the disks down, but caused no shutdown.

**Comment by ngandrass** [comment]
Thanks for further investigating :)

> And using -u zpool instead is a usable option, but it would be worth emphasizing the possible issue in the user instructions.

Yes, when you want to use zpool names instead of device names you need to switch from `-u disk` (default) to `-u zpool`, as described in the help text (`-h`) and `README`. 

Regarding the device identifier problem: I could add a note about this to the `README` file. However, nowadays most people seem to use SCALE anyway and therefore potentially face this problem. Wouldn't it be better to make the zpool mode the default, replacing the "fragile" disk device identifiers?

> Yes, the Debian shutdown command does not recognize the -p argument as FreeBSD does. The same result (Halt or Power off after shutdown) can be achieved with -h under Debian based systems. The script would somehow need to distinguish between these two.

I should be able to distinguish between both somehow. The script already does this for `camcontrol`/`hdparm`.

**Comment by SayHeading** [comment]
...so, back up again...

> Side question because I have problems with that: If a drive is spun down for you, does it stay that way for a prolonged period of time without waking up "on its own"?

The disks stayed spun down for the last few hours (only manually/acoustically confirmed). Scale Cobia is a fresh vanilla install without anything else running beyond the shutdown_script and one defined share which of course wasn't used.

**Comment by Salvoxia** [comment]
> The disks stayed spun down for the last few hours (only manually/acoustically confirmed). Scale Cobia is a fresh vanilla install without anything else running beyond the shutdown_script and one defined share which of course wasn't used.

Thanks for reporting back.
Just to be sure: Do you have the S.M.A.R.T. service enabled and running in `System Settings-> Services`, and if so, is S.M.A.R.T. enabled for the disks you tested spindown with?
My disks keep waking up at latest 30s after spindown, which I guess is a good way to an early grave...

**Comment by SayHeading** [comment]
Yes, it is enabled and running (also SMB and SSH), all 3 disks that shall spin down have the same settings:
![Bildschirmfoto 2024-01-15 um 19 45 33](https://github.com/ngandrass/truenas-spindown-timer/assets/156519194/011e578b-1143-4a62-9f39-55a0312162dc)

**Comment by SayHeading** [comment]
Update for Version 2.3.0 on SCALE 24.10.1:

First of all, with one 24.x-ish version of Scale, it is no longer possible to start scripts from the boot-pool. I had to move spindown_timer.sh to a different dataset.
But again with version 2.3.0 I have to replace the "-p" with "-h" command within the script in line 619, otherwise my system would not should down.

**Comment by ngandrass** [comment]
Yeah, the `-p` version seems to be a TrueNAS CORE / FreeBSD specific. I'd need to implement a way to distinguish between both systems. Should not be too hard ...

**Comment by ngandrass** [comment]
Yeah, there seems to be a difference between the arguments of the FreeBSD (TrueNAS CORE) and the Linux (TrueNAS SCALE) version of `shutdown`. Therefore we'd need to differentiate here. The easiest way would probably be to check for the existence of `/etc/debian_version` to know which version of TrueNAS we are running on.

The problem is, that I'm currently heavily occupied by other tasks and currently struggle maintaining this repo a little ... I hope that I can get some of this done within the next days.

**Comment by ngandrass** [comment]
@SayHeading I pushed an updated version of the script on the `develop` branch: https://github.com/ngandrass/truenas-spindown-timer/blob/develop/spindown_timer.sh

Can you please try if this solves your problem? Thanks :)

**Comment by SayHeading** [comment]
Moin Nils,

ich habe zwar mein „Problem“ schon durch händischen Austausch im Code der 2.0.3 gelöst,
werde das aber gern morgen mit der aktualisierten Version ausprobieren.
Die Rückmeldung gibt es dann über github (auch…).

Vielen Dank und einen entspannten Restsonntag,

Stefan

**Comment by ngandrass** [comment]
Thanks for testing! I'll wait a few days for #45 and release a new version after it :)

**Comment by ngandrass** [comment]
Version 2.4.0 is now officially released. You can use the `-x smartctl` argument to force use of `smartctl` instead of `hdparm` which should fix your issue :)
